src/main/java/org/nuxeo/wopi/jaxrs/FilesEndpoint.java

```diff
@@ -0,0 +1,613 @@
+    @POST
+    @Path("contents")
+    public Object doPostContents(@HeaderParam(OVERRIDE) Operation operation) {
+        if (Operation.PUT.equals(operation)) {
+            return putFile();
+        }
+        throw new BadRequestException();
+    }
+
+    /**
+     * Implements the PutFile operation.
+     * <p>
+     * See <a href="https://wopi.readthedocs.io/projects/wopirest/en/latest/files/PutFile.html"></a>.
+     */
+    public Object putFile() {
+        checkWritePropertiesPermission();
+
+        if (!doc.isLocked()) {
+            if (blob.getLength() == 0) {
+                return updateBlob();
+            }
+            response.addHeader(LOCK, "");
+            throw new ConflictException();
+        }
+
+        String currentLock = getCurrentLock();
+        String lock = getHeader(LOCK);
+        if (lock.equals(currentLock)) {
+            return updateBlob();
+        }
+
+        return buildConflictResponse(currentLock);
+    }
+
+    /**
+     * Updates the document's blob from a new one.
+     *
+     * @return the expected response for the PutFile operation, with the 'X-WOPI-ItemVersion' header set.
+     */
+    protected Response updateBlob() {
+        Blob newBlob = createBlobFromRequestBody(blob.getFilename(), blob.getMimeType());
+        doc.setPropertyValue(FILE_CONTENT_PROPERTY, (Serializable) newBlob);
+        doc.putContextData(SOURCE, WOPI_SOURCE);
+        doc = session.saveDocument(doc);
+        response.addHeader(ITEM_VERSION, doc.getVersionLabel());
+        return Response.ok().build();
+    }
+
+    /**
+     * Creates a new blob from the request body, given a {@code filename} and an optional {@code mimeType}.
+     *
+     * @return the new blob
+     */
+    protected Blob createBlobFromRequestBody(String filename, String mimeType) {
+        try (InputStream is = request.getInputStream()) {
+            Blob newBlob = Blobs.createBlob(is);
+            newBlob.setFilename(filename);
+            if (mimeType != null) {
+                newBlob.setMimeType(mimeType);
+            }
+            return newBlob;
+        } catch (IOException e) {
+            throw new NuxeoException(e);
+        }
+    }
+
+    /**
+     * Implements the Unlock operation.
+     * <p>
+     * See <a href="https://wopirest.readthedocs.io/en/latest/files/Unlock.html"></a>.
+     */
+    protected Object unlock() {
+        String lock = getHeader(LOCK);
+        return unlockOrRefresh(lock, true);
+    }
+
+    /**
+     * Implements the RefreshLock operation.
+     * <p>
+     * See <a href="https://wopirest.readthedocs.io/en/latest/files/RefreshLock.html"></a>.
+     */
+    protected Object refreshLock() {
+        String lock = getHeader(LOCK);
+        return unlockOrRefresh(lock, false);
+    }
+
+    protected int getMaxExpectedSize(String maxExpectedSizeHeader) {
+        if (!StringUtils.isBlank(maxExpectedSizeHeader)) {
+            try {
+                return Integer.parseInt(maxExpectedSizeHeader, 10);
```

Comment:
You should use a `long` here. And base 10 is the default.

Reply:
Even though the header is an integer as specified by the WOPI protocol?
See https://wopi.readthedocs.io/projects/wopirest/en/latest/files/GetFile.html, **Request Headers**.

Reply:
The `X-WOPI-MaxExpectedSize` is an integer according to WOPI spec.